### ISSUE TITLE
grpc-1.21.3: plugins don't link when protobuf is built statically and lite mode is used

### ISSUE THREAD

**Issue by BigBigos** [grpc/grpc#19472]
When protobuf is built statically and I try to then build gRPC in lite mode (`-DgRPC_USE_PROTO_LITE=On`), any plugin doesn't link:

```
: && /usr/bin/c++  -std=c++11 -O2 -g -DNDEBUG  -rdynamic CMakeFiles/grpc_ruby_plugin.dir/src/compiler/ruby_plugin.cc.o  -o grpc_ruby_plugin  /home/user/packages/etcd/lib/libprotoc.a /home/user/packages/etcd/lib/libprotobuf-lite.a -ldl -lrt -lm -lpthread libgrpc_plugin_support.a /home/user/packages/etcd/lib/libprotoc.a /home/user/packages/etcd/lib/libprotobuf-lite.a -ldl -lrt -lm -lpthread && :
/usr/bin/ld: /home/user/packages/etcd/lib/libprotoc.a(plugin.o): in function `google::protobuf::compiler::GenerateCode(google::protobuf::compiler::CodeGeneratorRequest const&, google::protobuf::compiler::CodeGenerator const&, google::protobuf::compiler::CodeGeneratorResponse*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*)':
/home/user/src/protobuf-3.8.0/build/src/../../src/google/protobuf/compiler/plugin.cc:106: undefined reference to `google::protobuf::DescriptorPool::DescriptorPool()'
...
```

The reason is that `libprotoc` requires `libprotobuf` (not the `-lite` version) but gRPC links everything against `libprotobuf-lite` when in the `gRPC_USE_PROTO_LITE` mode.

I have worked it around by adding `libprotobuf` to `_gRPC_PROTOBUF_PROTOC_LIBRARIES` in `cmake/protobuf.cmake`:

```
diff -ru grpc-1.21.3.orig/cmake/protobuf.cmake grpc-1.21.3.patched/cmake/protobuf.cmake
--- grpc-1.21.3.orig/cmake/protobuf.cmake	2019-06-04 00:23:18.000000000 +0200
+++ grpc-1.21.3.patched/cmake/protobuf.cmake	2019-06-25 10:51:10.227672601 +0200
@@ -65,7 +65,7 @@
       set(_gRPC_PROTOBUF_LIBRARIES ${PROTOBUF_LIBRARIES})
     endif()
     if(TARGET protobuf::libprotoc)
-      set(_gRPC_PROTOBUF_PROTOC_LIBRARIES protobuf::libprotoc)
+      set(_gRPC_PROTOBUF_PROTOC_LIBRARIES protobuf::libprotoc protobuf::libprotobuf) # libprotoc requires full libprotobuf
       # extract the include dir from target's properties
       get_target_property(_gRPC_PROTOBUF_WELLKNOWN_INCLUDE_DIR protobuf::libprotoc INTERFACE_INCLUDE_DIRECTORIES)
     else()
```

**Comment by a11r** [comment]
It is not clear if it is worth supporting lite mode here. Requesting AspirinSJL to investigate.

**Comment by jtattermusch** [comment]
I'd say this is worth doing only if it's an easy fix and doesn't require a major overhaul.

**Comment by zackgalbreath** [comment]
I was not able to reproduce the linker errors described in this issue. However, building gRPC with `gRPC_USE_PROTO_LITE` did cause compile errors for me:

```
/.../grpc/src/src/cpp/server/channelz/channelz_service.cc:31:78: error:
invalid conversion from ‘grpc::protobuf::Message* {aka google::protobuf::MessageLite*}’ to ‘google::protobuf::Message*’ [-fpermissive]
```

This issue is addressed in #21011

**Comment by jtattermusch** [comment]
Looks like the original problem cannot be reproduced. The channelz_service build problem will be handled separately.
